AsyncDisplayKit/ASVideoNode.mm

```diff
@@ -0,0 +1,276 @@
+
+
+#import "ASVideoNode.h"
+
+@interface ASVideoNode () {
+  ASDN::RecursiveMutex _lock;
+  
+  __weak id<ASVideoNodeDatasource> _datasource;
+  
+  BOOL _shouldBePlaying;
+  AVAsset *_asset;
+  AVPlayerItem *_currentItem;
+  ASButtonNode *_playButton;
+  ASDisplayNode *_playerNode;
+  ASDisplayNode *_spinner;
+}
+
+@end
+
+@interface ASDisplayNode ()
+- (void)setInterfaceState:(ASInterfaceState)newState;
+@end
+
+@implementation ASVideoNode
+
+- (instancetype)init {
+  if (!(self = [super init])) { return nil; }
+  
+  _playerNode = [[ASDisplayNode alloc] initWithLayerBlock:^CALayer *{
+    AVPlayerLayer *playerLayer = [[AVPlayerLayer alloc] init];
+    playerLayer.player = [[AVPlayer alloc] init];
+    return playerLayer;
+  }];
+  [self addSubnode:_playerNode];
+  
+  self.gravity = ASVideoGravityResizeAspect;
+  
+  [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(restartVideo:) name:AVPlayerItemDidPlayToEndTimeNotification object:nil];
+  
+  return self;
+}
+
+- (void)setInterfaceState:(ASInterfaceState)newState
+{
+  [super setInterfaceState:newState];
+  
+  if (!(newState & ASInterfaceStateVisible)) {
+    [self pause];
+    [_spinner removeFromSupernode];
+  } else {
+    if (_shouldBePlaying) {
+      [self play];
+    }
+    if (_spinner) {
+      [self addSubnode:_spinner];
+    }
+  }
+}
+
+- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context
+{
+  if ([[change objectForKey:@"new"] integerValue] == AVPlayerItemStatusReadyToPlay) {
+    if ([self.subnodes containsObject:_spinner]) {
+      [_spinner removeFromSupernode];
+      _spinner = nil;
+    }
+  }
+  
+  if ([[change objectForKey:@"new"] integerValue] == AVPlayerItemStatusFailed) {
+    
+  }
+}
+
+- (void)restartVideo:(NSNotification *)notification
+{
+  if ( [[[notification object] asset] isEqual:_asset]) {
+    [[((AVPlayerLayer *)_playerNode.layer) player] seekToTime:CMTimeMakeWithSeconds(0, 1)];
+    
+    if (_autorepeat) {
+      [self play];
+    } else {
+      [self pause];
+    }
+  }
+}
+
+- (void)layoutDidFinish
+{
+  _playerNode.frame = self.bounds;
+}
+
+- (void)didLoad {
+  [super didLoad];
+  
+  UITapGestureRecognizer *tap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(tapped)];
```

Comment:
Actually let's delete -didLoad because we should avoid creating a gesture recognizer if at all possible, which I think we can do by making sure this node is an ASControlNode, and adding a target / action directly to ourselves in init which calls -tapped.

We need to make sure the ASButtonNode for the play button can also trigger this, though.  Probably also want the play button to look selected even if the user taps somewhere else on the video.  So it might be best to set userInteractionEnabled = NO on the play button, and then even forward the -touchesBegan / ended methods?  hmm...maybe override hitTest: on this node to always hitTest: to the playButton, and then allow it to have user interaction, and set the target / action there would be best.

In any case, you should be able to kill the recognizer √